<a href="https://colab.research.google.com/github/BlairRong/DeleteMe/blob/main/3.26_lab2_OCR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Install Required Libraries

In [ ]:
# Install Tesseract OCR engine and Python packages
!apt-get install -y tesseract-ocr
!pip install pytesseract pillow

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 6 not upgraded.


Import Libraries and Upload Image

In [ ]:
import pytesseract
from PIL import Image
from google.colab import files

# Upload an image file
print("Please upload an image containing text:")
uploaded = files.upload()
filename = next(iter(uploaded))
print(f"\nFile uploaded: {filename}")

Please upload an image containing text:


Saving receipt.png to receipt (1).png

File uploaded: receipt (1).png


Task 1:  Extract Text, Store, and Split into Lines

In [ ]:
# Open the image
img = Image.open(filename)

# Extract text using Tesseract
extracted_text = pytesseract.image_to_string(img)

# IMPORTANT: Split the text into lines (removing empty lines)
lines = [line.strip() for line in extracted_text.splitlines() if line.strip()]

# Print the results
print("\n--- Extracted Text ---")
print(extracted_text)

print("\n--- Stored in variable 'extracted_text' ---")
print("Text variable length:", len(extracted_text))

print("\n--- Split into lines ---")
for i, line in enumerate(lines, 1):
    print(f"{i}: {line}")


--- Extracted Text ---
Walmar ><.

Save money. Live better.

(3
MANAG:
231
NEW PHI
ST# 02115 OP#
PET TOY
FLOPPY PUPPY
SSSUPREME S
2.5 SQUEAK
MUNCHY DMBEL
DOG TREAT
PED PCH 1
PED PCH 1
COUPON 23100
HNYMD SMORES
FRENCH DRSNG
3 ORANGES
BABY CARROTS
COLLARDS
CALZONE
MM RVW MNT
STKOBRLPLABL
STKOBRLPLABL
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
BLING BEADS
GREAT VALUE
LIPTON
DRY DOG

TAX

US DEBIT
APPROVAL # 57
REF # 7209005
TRANS ID - 38
VALIDATION -
PAYMENT SERVI!

AID A00000009
TC 51319CA81D
TERMINAL # SC
*Signature Ve

07/28

#
TCH 0443

Low Prices Y¥

 

 

 

 

 

 

 

 

 

 

 

 

 

 

 

 

07/28

#eEC

30 ) 339 - 3991
ER DIANA EARNEST
BLUEBELL DR SW

LADELPHIA OH 44663
009044 TE# 44 TR# 01301
004747571658 1.97 X
004747514846 1.97 X
070060332153 4.97 X
084699803238 5.92 X
068113108796 3.77 X
007119013654 2.92 xX
002310011802 0.50 X
002310011802 0.50 X
052310037000 1.00-0
088491226837 F 3.98 0
004132100655 F 1.98 0
001466835001 ® 5.47 N
003338366602 1.48 N
00000000461

Q1:
Why is it useful to structure the text instead of just printing it?

A1:
Structuring the text (e.g., splitting into lines, storing in a variable) makes it possible to **programmatically search, filter, and extract specific information** without manual scanning.  

- Ican isolate a single line (e.g., `lines[83]` contains `"TOTAL 98.21"`).  
- I can loop through lines to find keywords like `"TOTAL"`, `"SUBTOTAL"`, or `"TAX"` and then parse the corresponding values.  
- This enables **automation** – I can build a script that calculates totals, validates amounts, or extracts dates without human intervention.  

Simply printing the raw text gives a human‑readable output, but structured data unlocks further processing and integration into larger systems (e.g., accounting software, inventory management).

Task 2: Add extraction functions



In [ ]:
# Task 2 – Extract specific information (numbers, total, date)

import re

def extract_numbers(text):
    """Return all numbers (integers and decimals) found in the text."""
    return re.findall(r'\d+\.?\d*', text)

def find_total(lines):
    """Look for a line containing 'TOTAL' and extract the number after it."""
    for line in lines:
        if 'TOTAL' in line.upper():
            # Try to find the first number in that line
            numbers = extract_numbers(line)
            if numbers:
                return numbers[0]  # return the first number (assumed total)
    return None

def find_date(lines):
    """Look for a line containing a date pattern (MM/DD or MM/DD/YY)."""
    date_pattern = r'\b\d{1,2}/\d{1,2}(?:/\d{2,4})?\b'
    for line in lines:
        matches = re.findall(date_pattern, line)
        if matches:
            return matches[0]
    return None

Integrate into the main script

In [ ]:
# After extracting and splitting lines (from Task 1)

all_numbers = extract_numbers(extracted_text)
total_amount = find_total(lines)
date_found = find_date(lines)

print("\n--- All extracted numbers ---")
print(all_numbers)

print("\n--- Total amount found ---")
print(total_amount if total_amount else "Not found")

print("\n--- Date found ---")
print(date_found if date_found else "Not found")


--- All extracted numbers ---
['3', '231', '02115', '2.5', '1', '1', '23100', '3', '57', '7209005', '38', '00000009', '51319', '81', '07', '28', '0443', '07', '28', '30', '339', '3991', '44663', '009044', '44', '01301', '004747571658', '1.97', '004747514846', '1.97', '070060332153', '4.97', '084699803238', '5.92', '068113108796', '3.77', '007119013654', '2.92', '002310011802', '0.50', '002310011802', '0.50', '052310037000', '1.00', '0', '088491226837', '3.98', '0', '004132100655', '1.98', '0', '001466835001', '5.47', '003338366602', '1.48', '00000000461', '4', '7', '1.24', '005208362080', '2.50', '0', '003399105848', '19.77', '001558679414', '1.97', '001558679414', '1.97', '001558679410', '0.97', '001558679410', '0.97', '001558679410', '0.97', '001558679410', '0.97', '076594060699', '0.97', '007874203191', '9.97', '0', '001200011224', '4.48', '002310011035', '12.44', '93.62', '1', '6.750', '4.59', '98.21', '98.21', '0166', '2868', '44961', '7209239650894', '87', '80840', '22', '7', '0

Answer:
**Did it work reliably?**  
Partly. It successfully extracted numbers and found a date and a value that appears to be a total, but it picked **SUBTOTAL** (93.62) instead of the actual **TOTAL** (98.21). The date extraction returned `07/28`, which exists on the receipt, but there are also other date‑like strings (e.g., `/17 02:39:48`).  



**What problems did you encounter?**  
- The keyword `"TOTAL"` matched the first occurrence – `SUBTOTAL` – rather than the final total line.  
- The date regex matched the first `MM/DD` pattern it found, which may not be the correct one if multiple dates appear.  
- The extracted numbers list is long and includes many irrelevant numbers (e.g., product codes, reference numbers) that are not useful for total extraction.  
- OCR errors (e.g., `"TOTAL"` being misread as `"TOT AL"` or missing entirely) would break the extraction.  

**In short:** the approach works for clean, well‑structured receipts, but it needs more robust logic (e.g., searching for `"TOTAL"` after `"SUBTOTAL"`, or looking for a line with both `"TOTAL"` and a decimal number) to be reliable across different layouts.



答：它运行可靠吗？

部分可靠。它成功提取了数字，并找到了一个日期和一个看似总计的值，但它提取的是小计 (93.62) 而不是实际的总计 (98.21)。日期提取结果为 07/28，该日期存在于收据上，但还有其他类似日期的字符串（例如，/17 02:39:48）。

您遇到了哪些问题？

关键字“总计”匹配的是第一个出现的“小计”，而不是最终的总计行。

日期正则表达式匹配的是它找到的第一个 MM/DD 模式，如果出现多个日期，则该模式可能不正确。

提取的数字列表很长，其中包含许多与总计提取无关的数字（例如，产品代码、参考编号）。

OCR 错误（例如，“总计”被误读为“TOT AL”或完全缺失）会导致提取失败。

简而言之：这种方法适用于干净、结构良好的收据，但需要更强大的逻辑（例如，在“小计”之后搜索“总计”，或查找同时包含“总计”和小数的行）才能在不同的布局中保持可靠性。

Task 3 – Build a Simple OCR Application (Receipt Analyzer)

I will create a small program that:

Uploads an image (e.g., a receipt) in Google Colab.

Runs OCR to extract all text.

Prints the full extracted text.

Prints a list of all numbers found (key values often used in analysis).


In [ ]:
# Task 3 – Simple OCR Application (Receipt Analyzer)

# Install dependencies
!apt-get install -y tesseract-ocr
!pip install pytesseract pillow

import pytesseract
import re
from PIL import Image
from google.colab import files

# Upload image
print("Please upload an image (e.g., a receipt):")
uploaded = files.upload()
filename = next(iter(uploaded))
print(f"\nFile uploaded: {filename}")

# OCR
img = Image.open(filename)
extracted_text = pytesseract.image_to_string(img)

# Extract numbers
numbers = re.findall(r'\d+\.?\d*', extracted_text)

# Output
print("\n" + "="*50)
print("FULL EXTRACTED TEXT")
print("="*50)
print(extracted_text)

print("\n" + "="*50)
print("EXTRACTED KEY VALUES (numbers)")
print("="*50)
if numbers:
    print(numbers)
else:
    print("No numbers found.")

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 6 not upgraded.
Please upload an image (e.g., a receipt):


Saving receipt.png to receipt (2).png

File uploaded: receipt (2).png

FULL EXTRACTED TEXT
Walmar ><.

Save money. Live better.

(3
MANAG:
231
NEW PHI
ST# 02115 OP#
PET TOY
FLOPPY PUPPY
SSSUPREME S
2.5 SQUEAK
MUNCHY DMBEL
DOG TREAT
PED PCH 1
PED PCH 1
COUPON 23100
HNYMD SMORES
FRENCH DRSNG
3 ORANGES
BABY CARROTS
COLLARDS
CALZONE
MM RVW MNT
STKOBRLPLABL
STKOBRLPLABL
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
BLING BEADS
GREAT VALUE
LIPTON
DRY DOG

TAX

US DEBIT
APPROVAL # 57
REF # 7209005
TRANS ID - 38
VALIDATION -
PAYMENT SERVI!

AID A00000009
TC 51319CA81D
TERMINAL # SC
*Signature Ve

07/28

#
TCH 0443

Low Prices Y¥

 

 

 

 

 

 

 

 

 

 

 

 

 

 

 

 

07/28

#eEC

30 ) 339 - 3991
ER DIANA EARNEST
BLUEBELL DR SW

LADELPHIA OH 44663
009044 TE# 44 TR# 01301
004747571658 1.97 X
004747514846 1.97 X
070060332153 4.97 X
084699803238 5.92 X
068113108796 3.77 X
007119013654 2.92 xX
002310011802 0.50 X
002310011802 0.50 X
052310037000 1.00-0
088491226837 F 3.98 0
00413210

Task 4 – Improve with Prompting (Text Cleaning & Readability)



Step 1 – Add a cleaning function


The function will:

Remove excessive newlines.

Strip leading/trailing spaces from each line.

Filter out lines that consist only of symbols (e.g., ><, ---).

Replace multiple spaces with a single space.

Optionally remove stray characters that are not letters, numbers, or common punctuation.

In [ ]:
import re

def clean_text(text):
    """Clean OCR output: normalize spaces, remove noise, and filter empty lines."""
    # Replace multiple newlines with a single newline
    text = re.sub(r'\n\s*\n', '\n', text)
    # Remove lines that are mostly non‑alphanumeric (e.g., decoration lines)
    lines = []
    for line in text.splitlines():
        # Remove leading/trailing spaces
        line = line.strip()
        # Keep only lines with at least one alphanumeric character
        if re.search(r'[a-zA-Z0-9]', line):
            # Replace multiple spaces inside the line
            line = re.sub(r' +', ' ', line)
            lines.append(line)
    return '\n'.join(lines)

Step 2 – Integrate cleaning into the OCR app


I modify the previous script to call clean_text() before printing the full text and extracting numbers.

In [ ]:
# Task 4 – Improved OCR with text cleaning

!apt-get install -y tesseract-ocr
!pip install pytesseract pillow

import pytesseract
import re
from PIL import Image
from google.colab import files

def clean_text(text):
    """Clean OCR output: normalize spaces, remove noise, filter empty lines."""
    text = re.sub(r'\n\s*\n', '\n', text)
    lines = []
    for line in text.splitlines():
        line = line.strip()
        if re.search(r'[a-zA-Z0-9]', line):
            line = re.sub(r' +', ' ', line)
            lines.append(line)
    return '\n'.join(lines)

print("Upload an image (e.g., receipt):")
uploaded = files.upload()
filename = next(iter(uploaded))
print(f"\nFile: {filename}")

# OCR
img = Image.open(filename)
raw_text = pytesseract.image_to_string(img)

# Clean text
cleaned_text = clean_text(raw_text)

# Extract numbers from cleaned text (still useful)
numbers = re.findall(r'\d+\.?\d*', cleaned_text)

# Output
print("\n" + "="*50)
print("CLEANED EXTRACTED TEXT")
print("="*50)
print(cleaned_text)

print("\n" + "="*50)
print("EXTRACTED NUMBERS")
print("="*50)
print(numbers if numbers else "No numbers found.")

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 6 not upgraded.
Upload an image (e.g., receipt):


Saving receipt.png to receipt (3).png

File: receipt (3).png

CLEANED EXTRACTED TEXT
Walmar ><.
Save money. Live better.
(3
MANAG:
231
NEW PHI
ST# 02115 OP#
PET TOY
FLOPPY PUPPY
SSSUPREME S
2.5 SQUEAK
MUNCHY DMBEL
DOG TREAT
PED PCH 1
PED PCH 1
COUPON 23100
HNYMD SMORES
FRENCH DRSNG
3 ORANGES
BABY CARROTS
COLLARDS
CALZONE
MM RVW MNT
STKOBRLPLABL
STKOBRLPLABL
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
BLING BEADS
GREAT VALUE
LIPTON
DRY DOG
TAX
US DEBIT
APPROVAL # 57
REF # 7209005
TRANS ID - 38
VALIDATION -
PAYMENT SERVI!
AID A00000009
TC 51319CA81D
TERMINAL # SC
*Signature Ve
07/28
TCH 0443
Low Prices Y¥
07/28
#eEC
30 ) 339 - 3991
ER DIANA EARNEST
BLUEBELL DR SW
LADELPHIA OH 44663
009044 TE# 44 TR# 01301
004747571658 1.97 X
004747514846 1.97 X
070060332153 4.97 X
084699803238 5.92 X
068113108796 3.77 X
007119013654 2.92 xX
002310011802 0.50 X
002310011802 0.50 X
052310037000 1.00-0
088491226837 F 3.98 0
004132100655 F 1.98 0
001466835001 ® 5.47 N
003338366602 1.48 N
00000000461 

Task4

Q1: How did the output change?

Empty lines were removed, making the text more compact.

Lines consisting only of symbols (e.g., ><, ---) were filtered out.

The text became cleaner and easier to scan manually.

Q2: Was it easier to use?

Yes. The cleaned text is more readable and ready for further processing (e.g., searching for totals, dates) because it contains fewer visual artifacts and is more structured. Removing noise also improves the accuracy of any subsequent regex searches.

Task 5 – Add Simple Logic (Expense Categorization)

I extend the OCR application to automatically detect the total amount from a receipt and classify it as High expense (> 500) or Normal expense (≤ 500).

In [ ]:
Step 1 – Create a function to extract the total amount

I look for a line containing the word TOTAL (or TOT, Total) and then extract the first number in that line.

In [ ]:
def extract_total(text):
    """
    Extract the total amount from OCR text.
    Looks for lines containing 'TOTAL', then extracts the first number.
    Returns the total as a float, or None if not found.
    """
    for line in text.splitlines():
        if 'TOTAL' in line.upper():
            numbers = re.findall(r'\d+\.?\d*', line)
            if numbers:
                return float(numbers[0])
    return None

Step 2 – Add the logic to the main program

After cleaning the text and extracting numbers, I call extract_total() and print the category.

In [ ]:
# Task 5 – OCR with expense categorization

!apt-get install -y tesseract-ocr
!pip install pytesseract pillow

import pytesseract
import re
from PIL import Image
from google.colab import files

def clean_text(text):
    """Clean OCR output: normalize spaces, remove noise, filter empty lines."""
    text = re.sub(r'\n\s*\n', '\n', text)
    lines = []
    for line in text.splitlines():
        line = line.strip()
        if re.search(r'[a-zA-Z0-9]', line):
            line = re.sub(r' +', ' ', line)
            lines.append(line)
    return '\n'.join(lines)

def extract_total(text):
    """Extract the total amount (first number in a line containing 'TOTAL')."""
    for line in text.splitlines():
        if 'TOTAL' in line.upper():
            numbers = re.findall(r'\d+\.?\d*', line)
            if numbers:
                return float(numbers[0])
    return None

# Upload image
print("Please upload an image of a receipt:")
uploaded = files.upload()
filename = next(iter(uploaded))
print(f"\nFile uploaded: {filename}")

# OCR
img = Image.open(filename)
raw_text = pytesseract.image_to_string(img)

# Clean text
cleaned_text = clean_text(raw_text)

# Extract total
total = extract_total(cleaned_text)

# Output
print("\n" + "="*50)
print("CLEANED EXTRACTED TEXT")
print("="*50)
print(cleaned_text)

print("\n" + "="*50)
print("EXPENSE CATEGORY")
print("="*50)
if total is None:
    print("Could not find total amount.")
else:
    print(f"Total: {total:.2f}")
    if total > 500:
        print("High expense")
    else:
        print("Normal expense")

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 6 not upgraded.
Please upload an image of a receipt:


Saving receipt.png to receipt (4).png

File uploaded: receipt (4).png

CLEANED EXTRACTED TEXT
Walmar ><.
Save money. Live better.
(3
MANAG:
231
NEW PHI
ST# 02115 OP#
PET TOY
FLOPPY PUPPY
SSSUPREME S
2.5 SQUEAK
MUNCHY DMBEL
DOG TREAT
PED PCH 1
PED PCH 1
COUPON 23100
HNYMD SMORES
FRENCH DRSNG
3 ORANGES
BABY CARROTS
COLLARDS
CALZONE
MM RVW MNT
STKOBRLPLABL
STKOBRLPLABL
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
STKO SUNFLWR
BLING BEADS
GREAT VALUE
LIPTON
DRY DOG
TAX
US DEBIT
APPROVAL # 57
REF # 7209005
TRANS ID - 38
VALIDATION -
PAYMENT SERVI!
AID A00000009
TC 51319CA81D
TERMINAL # SC
*Signature Ve
07/28
TCH 0443
Low Prices Y¥
07/28
#eEC
30 ) 339 - 3991
ER DIANA EARNEST
BLUEBELL DR SW
LADELPHIA OH 44663
009044 TE# 44 TR# 01301
004747571658 1.97 X
004747514846 1.97 X
070060332153 4.97 X
084699803238 5.92 X
068113108796 3.77 X
007119013654 2.92 xX
002310011802 0.50 X
002310011802 0.50 X
052310037000 1.00-0
088491226837 F 3.98 0
004132100655 F 1.98 0
001466835001 ® 5.47 N
003338366602 1.48 N
000

Why this is useful?

Adding simple logic turns raw OCR into actionable information. Instead of just extracting text, the system can now help users quickly categorize expenses, which is a practical step toward automation.

Task6: Reflection


**Where does OCR stop and “understanding” begin?**  
OCR ends when it converts visual text into machine‑readable strings. Understanding begins when we interpret that text – e.g., identifying which number is the total, what the date means, or that “TOTAL” signals the final amount. The separation is between **extraction** (OCR) and **interpretation** (parsing, context awareness).

**Did my system actually understand the receipt?**  
No, it did not understand. It simply applied rules: it looked for a line containing the word “TOTAL” and took the first number. It has no concept of what a receipt is, what subtotal vs. total means, or that “TOTAL” might be misread. It performed pattern matching, not understanding.

**What would be needed to make it smarter?**  
- **Contextual knowledge**: Know that receipts often have “SUBTOTAL” followed by “TOTAL”, so it should look for the second occurrence.  
- **Better parsing**: Use more robust regex or a small ML model to identify key fields (total, date, tax) based on layout and proximity.  
- **Fallback mechanisms**: Handle OCR errors (e.g., “T0TAL” or “TOT AL”) using fuzzy matching.  
- **Document structure awareness**: Distinguish between line items, headers, and summary sections.  
- **Domain adaptation**: Train a small model on receipt layouts, or use a pre‑built service like Azure Document Intelligence that understands document structure.




**OCR 的终点在哪里？“理解”的起点又在哪里？** OCR 的终点是将视觉文本转换为机器可读的字符串。理解的起点在于我们对这些文本进行解释——例如，识别哪个数字是总计，日期的含义，或者“总计”表示最终金额。区别在于**提取**（OCR）和**解释**（解析、上下文感知）。

**我的系统真的理解了收据吗？** 没有，它并没有理解。它只是应用了规则：查找包含“总计”一词的行，并提取第一个数字。它并不理解收据是什么，小计和总计的含义是什么，或者“总计”可能会被误读。它执行的是模式匹配，而不是理解。

**如何才能让它更智能？**

- **上下文知识**：了解收据上通常会先出现“小计”，然后是“总计”，因此它应该查找第二个出现的“总计”。

- **更佳解析**：使用更强大的正则表达式或小型机器学习模型，根据布局和邻近性识别关键字段（总计、日期、税额）。

- **回退机制**：使用模糊匹配处理 OCR 错误（例如，“T0TAL”或“TOT AL”）。

- **文档结构感知**：区分行项目、标题和汇总部分。

- **领域自适应**：在收据布局上训练小型模型，或使用预构建的、能够理解文档结构的服务，例如 Azure 文档智能。

Task 7 — Advanced: Multi‑Image Receipt Analyzer with Field Extraction


In this task I extend the OCR system to:

Detect if an image contains a receipt (simple keyword‑based detection).

Extract specific fields – total amount and date.

Process multiple images in one run, showing results for each.

All code runs in Google Colab.

In [ ]:
# Task 7 – Advanced Multi‑Image Receipt Analyzer

# Install dependencies
!apt-get install -y tesseract-ocr
!pip install pytesseract pillow

import pytesseract
import re
from PIL import Image
from google.colab import files

def clean_text(text):
    """Remove extra whitespace and keep only lines with alphanumeric chars."""
    text = re.sub(r'\n\s*\n', '\n', text)
    lines = []
    for line in text.splitlines():
        line = line.strip()
        if re.search(r'[a-zA-Z0-9]', line):
            line = re.sub(r' +', ' ', line)
            lines.append(line)
    return '\n'.join(lines)

def is_receipt(text):
    """Detect if the text likely comes from a receipt."""
    keywords = ['TOTAL', 'SUBTOTAL', 'RECEIPT', 'TAX', 'CASH', 'CHANGE', 'ITEM']
    upper_text = text.upper()
    matches = sum(1 for kw in keywords if kw in upper_text)
    return matches >= 2

def extract_total(text):
    """Find total amount (looks for 'TOTAL' line and extracts first number)."""
    for line in text.splitlines():
        if 'TOTAL' in line.upper():
            numbers = re.findall(r'\d+\.?\d*', line)
            if numbers:
                return float(numbers[0])
    return None

def extract_date(text):
    """Find a date in common formats (MM/DD or MM/DD/YY)."""
    patterns = [r'\b\d{1,2}/\d{1,2}/\d{2,4}\b', r'\b\d{1,2}/\d{1,2}\b']
    for line in text.splitlines():
        for pat in patterns:
            match = re.search(pat, line)
            if match:
                return match.group()
    return None

def process_image(filename):
    img = Image.open(filename)
    raw = pytesseract.image_to_string(img)
    cleaned = clean_text(raw)
    return {
        'filename': filename,
        'receipt': is_receipt(cleaned),
        'total': extract_total(cleaned),
        'date': extract_date(cleaned),
        'preview': cleaned[:150].replace('\n', ' ')
    }

# Upload multiple images
print("Upload one or more images (hold Ctrl/Cmd to select multiple):")
uploaded = files.upload()
filenames = list(uploaded.keys())
print(f"\nUploaded {len(filenames)} file(s): {filenames}")

# Process each
results = [process_image(f) for f in filenames]

# Print summary
print("\n" + "="*60)
print("RECEIPT ANALYSIS SUMMARY")
print("="*60)
for r in results:
    print(f"\n📄 {r['filename']}")
    if r['receipt']:
        print("   ✅ Likely a receipt")
        print(f"   💰 Total: {r['total'] if r['total'] else 'not found'}")
        print(f"   📅 Date:  {r['date'] if r['date'] else 'not found'}")
    else:
        print("   ❌ Not a receipt (or keywords missing)")
    print(f"   👀 Preview: {r['preview']}")

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
tesseract-ocr is already the newest version (4.1.1-2.1build1).
0 upgraded, 0 newly installed, 0 to remove and 6 not upgraded.
Upload one or more images (hold Ctrl/Cmd to select multiple):


Saving handwriting.jpg to handwriting.jpg
Saving receipt.png to receipt (6).png

Uploaded 2 file(s): ['handwriting.jpg', 'receipt (6).png']

RECEIPT ANALYSIS SUMMARY

📄 handwriting.jpg
   ❌ Not a receipt (or keywords missing)
   👀 Preview: aT Lov XY Wh re SHRPPYY by 2 j2% 4 BORE. TOT ag LS vo —> ~-V/\M pe y TY! Or > . oe ARGO S PEP YL AWD FELON Ls | i ory OV) JO LOH PHN ey Wy) wy x shire

📄 receipt (6).png
   ✅ Likely a receipt
   💰 Total: 93.62
   📅 Date:  07/28
   👀 Preview: Walmar ><. Save money. Live better. (3 MANAG: 231 NEW PHI ST# 02115 OP# PET TOY FLOPPY PUPPY SSSUPREME S 2.5 SQUEAK MUNCHY DMBEL DOG TREAT PED PCH 1 P
